src/crimson/osd/osd_operation_sequencer.h

```diff
@@ -26,71 +25,139 @@ namespace crimson::osd {
       // ones, so that the following ops can be queued up here. we cannot let
       // the seastar scheduler to determine the order of performing these ops,
       // once they are unblocked after the first op of the same pg interval is
       // scheduled.
-      assert(prev_op > last_unblocked);
+      const auto prev_id = get_prev_id(op, registry);
+      assert(prev_id >= last_unblocked_id);
       handle.exit();
       ::crimson::get_logger(ceph_subsys_osd).debug(
-        "OpSequencer::start_op: {} waiting ({} > {})",
-        op, prev_op, last_unblocked);
-      have_green_light = unblocked.wait([prev_op, this] {
+        "OpSequencer::start_op: {} resequencing ({} >= {})",
+        op, prev_id, last_unblocked_id);
+      have_green_light = unblocked.wait([&op, &registry, this] {
         // wait until the previous op is unblocked
-        return last_unblocked == prev_op;
+        const bool unblocking =
+          get_prev_id(op, registry) == last_unblocked_id;
+        if (unblocking) {
+          // stop resequencing if everything is handled which means there is no
+          // operation after us. the range could be minimized by snapshotting
+          // `last_seen` on `maybe_reset()`.
+          // `<=` is to handle the situation when `last_seen` has finished out-
+          // of-the-order.
+          resequencing = !(last_seen_id <= op.get_id());
+        }
+        return unblocking;
       });
     }
-    return have_green_light.then([this_op, do_op=std::move(do_op), this]() mutable {
+    return have_green_light.then([&op, do_op=std::move(do_op), this]() mutable {
       auto result = seastar::futurize_invoke(std::move(do_op));
       // unblock the next one
-      last_unblocked = this_op;
+      last_unblocked_id = op.get_id();
       unblocked.broadcast();
       return result;
     });
   }
-  uint64_t get_last_issued() const {
-    return last_issued;
+  void finish_op_in_order(ClientRequest& op) {
+    ::crimson::get_logger(ceph_subsys_osd).debug(
+      "OpSequencer::{}: op={}, last_seen={}, last_unblocked={}, last_completed={}",
+      __func__, op.get_id(), last_seen_id, last_unblocked_id, last_completed_id);
+    assert(op.get_id() > last_completed_id);
+    last_completed_id = op.get_id();
+    op.finished = true;
   }
-  template <class OpT>
-  void finish_op(const OpT& op) {
-    assert(op.get_id() > last_completed);
-    last_completed = op.get_id();
+  void finish_op_out_of_order(ClientRequest& op,
+                              const OSDOperationRegistry& registry) {
+    ::crimson::get_logger(ceph_subsys_osd).debug(
+      "OpSequencer::{}: op={}, last_seen={}, last_unblocked={}, last_completed={}",
+      __func__, op.get_id(), last_seen_id, last_unblocked_id, last_completed_id);
+    // fix the last_unblocked_id. otherwise we wouldn't be able to leave
+    // the wait loop in start_op() as any concrete value of last_unblocked_id
+    // can wake at most one op and above we lowered this number to 0.
```

Comment:
i don't quite follow this line

> any concrete value of can wake at most one op and above we lowered this number to 0.

could you rephrase this sentence?

Reply:
A blocked operation waits on its predecessor. We can also say an op can wake its successor but only if there is any (the op isn't the very last one) – that's why "_at most one_".

However, this order is violated by `finish_op_out_of_order()`. By removing an op-in-the-middle we're also taking out the perspective-of-being-waken from its successor ("_lowered this number to 0_").

Will try rephrase.